filebeat/tests/integration/filestream_test.go

```diff
@@ -657,6 +671,471 @@ func TestFilestreamTakeOverFromLogInput(t *testing.T) {
+		if event.Has(fsnotify.Remove) {
+			t.Errorf("File %s should not have been removed, removal happened at %s",
+				event.Name,
+				time.Now().Format(time.RFC3339Nano))
+		}
+	})
+	fileWatcher.Start()
+	defer fileWatcher.Stop()
+
+	esURL := integration.GetESURL(t, "http")
+
+	// Create and start the proxy server
+	proxy := &DisablingProxy{target: &esURL, enabled: true}
+	server := &http.Server{
+		Addr:              "localhost:9201",
+		Handler:           proxy,
+		ReadHeaderTimeout: time.Second / 2,
+	}
+	go func() {
+		if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			t.Errorf("Proxy server failed: %s", err)
+		}
+	}()
+	defer server.Close()
+
+	proxyURL, err := url.Parse(server.Addr)
+	if err != nil {
+		t.Fatalf("cannot parse proxy URL: %s", err)
+	}
+
+	user := esURL.User.Username()
+	pass, _ := esURL.User.Password()
+	vars := map[string]any{
+		"homePath":    workDir,
+		"logfile":     logFile,
+		"testdata":    testDataPath,
+		"esHost":      proxyURL.String(),
+		"user":        user,
+		"pass":        pass,
+		"index":       index,
+		"gracePeriod": gracePeriod.String(),
+	}
+
+	cfgYAML := getConfig(t, vars, "delete", "real-es.yml")
+	filebeat.WriteConfigFile(cfgYAML)
+	filebeat.Start()
+
+	// Wait for data in ES
+	msgs := []string{}
+	require.Eventually(t, func() bool {
+		msgs = getEventsMsgFromES(t, index, 200)
+		return len(msgs) == len(logFileLines)/2
+	}, time.Second*10, time.Millisecond*100, "not all log messages have been found on ES")
+
+	// Wait for 1/2 of the grace period and add more data
+	time.Sleep(gracePeriod / 2)
+
+	// Add more data to the file
+	f, err := os.OpenFile(logFile, os.O_RDWR|os.O_APPEND, 0644)
+	if err != nil {
+		t.Fatalf("cannot open logfile to append data: %s", err)
+	}
+	logData2 := strings.Join(logFileLines[5:], "\n")
+	logData2 += "\n"
+	if _, err := f.WriteString(logData2); err != nil {
+		t.Fatalf("could not append data to log file: %s", err)
+	}
+	if err := f.Sync(); err != nil {
+		t.Fatalf("cannot flush log file: %s", err)
+	}
+	if err := f.Close(); err != nil {
+		t.Fatalf("cannot close log file: %s", err)
+	}
+
+	// Disable (aka block) the output
+	proxy.Disable()
+
+	// Wait twice the grace period before unblocking the output
+	blockedTimer := time.NewTimer(gracePeriod * 2)
+	<-blockedTimer.C
+
+	// Ensure log file still exists
+	if !fileExists(t, logFile) {
+		t.Fatal("file was removed while output was blocked")
+	}
+
+	// Unblock the output
+	proxy.Enable()
+
+	// Wait for the remaining data to be ingested
+	msgs = []string{}
+	require.Eventually(t, func() bool {
+		msgs = getEventsMsgFromES(t, index, 200)
+		return len(msgs) == len(logFileLines)
+	}, time.Second*10, time.Millisecond*100, "not all log messages have been found on ES")
+
+	dataShippedTs := time.Now()
+	fileRemovedChan := make(chan time.Time)
+	// All events have been found, allow file to be removed
+	// and get the removal timestamp
+	fileWatcher.SetEventCallback(func(event fsnotify.Event) {
```

Comment:
I think there is a race here because your query above is 10s but the grace period is 5s, so Filebeat is allowed to delete the file before the query above has succeeded but your logic here would fail the test in this situation. You only swap the deletion condition after the query succeeds.

Reply:
I see what you're saying. Those 10s there are very arbitrary, and usually the maximum wait time is quire generous because CI might be slow.

I only set the callback to get the deletion time after we assert the events are successfully ingested by ES, so if the file gets removed before we set this callback, the test will fail.

Anyways, I agree it makes sense to make it smaller than the grace period.